File: mobile/src/main/java/net/mm2d/dmsexplorer/view/view/ScrubBar.java
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.dmsexplorer.view.view;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.os.Build;
import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.ColorInt;
import android.support.annotation.Dimension;
import android.support.annotation.IntDef;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.MotionEvent;
import android.view.View;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * @author <a href="mailto:ryo@mm2d.net">大前良介 (OHMAE Ryosuke)</a>
 */
public class ScrubBar extends View {
    @Retention(RetentionPolicy.SOURCE)
    @IntDef({ACCURACY_NORMAL, ACCURACY_HALF, ACCURACY_QUARTER})
    public @interface Accuracy {
    }

    public static final int ACCURACY_NORMAL = 0;
    public static final int ACCURACY_HALF = 1;
    public static final int ACCURACY_QUARTER = 2;

    private static final int[] ACCURACY_RANKS = {
            ACCURACY_NORMAL, ACCURACY_HALF, ACCURACY_QUARTER,
    };
    private static final float[] ACCURACY = new float[]{
            1.1f, 0.5f, 0.25f,
    };
    private static final int RANK_MAX = ACCURACY_RANKS.length - 1;

    public interface ScrubBarListener {
        void onProgressChanged(
                @NonNull ScrubBar seekBar,
                int progress,
                boolean fromUser);

        void onStartTrackingTouch(@NonNull ScrubBar seekBar);

        void onStopTrackingTouch(@NonNull ScrubBar seekBar);

        void onAccuracyChanged(
                @NonNull ScrubBar seekBar,
                @Accuracy int accuracy);
    }

    private static final ScrubBarListener LISTENER = new ScrubBarListener() {
        @Override
        public void onProgressChanged(
                @NonNull final ScrubBar seekBar,
                final int progress,
                final boolean fromUser) {
        }

        @Override
        public void onStartTrackingTouch(@NonNull final ScrubBar seekBar) {
        }

        @Override
        public void onStopTrackingTouch(@NonNull final ScrubBar seekBar) {
        }

        @Override
        public void onAccuracyChanged(
                @NonNull final ScrubBar seekBar,
                @Accuracy final int accuracy) {
        }
    };

    @Dimension(unit = Dimension.DP)
    private static final int TRACK_WIDTH_DP = 3;
    @Dimension(unit = Dimension.DP)
    private static final int SMALL_THUMB_RADIUS_DP = 4;
    @Dimension(unit = Dimension.DP)
    private static final int LARGE_THUMB_RADIUS_DP = 8;
    @Dimension(unit = Dimension.DP)
    private static final int SCRUB_UNIT_LENGTH_DP = 150;
    @ColorInt
    private static final int DEFAULT_ENABLED_TRACK_COLOR = Color.argb(0x66, 0xff, 0xff, 0xff);
    @ColorInt
    private static final int DEFAULT_ENABLED_SECTION_COLOR = Color.argb(0xff, 0xff, 0xff, 0x0);
    @ColorInt
    private static final int DEFAULT_DISABLED_PROGRESS_COLOR = Color.argb(0xff, 0x80, 0x80, 0x80);
    @ColorInt
    private static final int DEFAULT_DISABLED_TRACK_COLOR = Color.argb(0x66, 0x60, 0x60, 0x60);
    @ColorInt
    private static final int DEFAULT_DISABLED_SECTION_COLOR = Color.argb(0x66, 0xff, 0xff, 0x0);

    @NonNull
    private final Paint mPaint;
    @NonNull
    private ScrubBarListener mScrubBarListener = LISTENER;
    @Dimension(unit = Dimension.PX)
    private final float mTrackWidth;
    @Dimension(unit = Dimension.PX)
    private final float mTrackWidthHalf;
    @Dimension(unit = Dimension.PX)
    private final float mSmallThumbRadius;
    @Dimension(unit = Dimension.PX)
    private final float mLargeThumbRadius;
    @Dimension(unit = Dimension.PX)
    private int mScrubUnitLength;
    @ColorInt
    private int mEnabledProgressColor;
    @ColorInt
    private int mEnabledTrackColor = DEFAULT_ENABLED_TRACK_COLOR;
    @ColorInt
    private int mEnabledSectionColor = DEFAULT_ENABLED_SECTION_COLOR;
    @ColorInt
    private int mDisabledProgressColor = DEFAULT_DISABLED_PROGRESS_COLOR;
    @ColorInt
    private int mDisabledTrackColor = DEFAULT_DISABLED_TRACK_COLOR;
    @ColorInt
    private int mDisabledSectionColor = DEFAULT_DISABLED_SECTION_COLOR;
    @ColorInt
    private int mProgressColor;
    @ColorInt
    private int mTrackColor;
    @ColorInt
    private int mSectionColor;
    @ColorInt
    private int mTopBackgroundColor;
    @ColorInt
    private int mBottomBackgroundColor;

    private int mProgress;
    private int mMax;
    @NonNull
    private List<Integer> mChapterList = Collections.emptyList();

    private int mAccuracyRank;
    private boolean mDragging;
    private float mStartX;
    private float mStartY;
    private int mBaseProgress;

    public ScrubBar(@NonNull final Context context) {
        this(context, null);
    }

    public ScrubBar(
            @NonNull final Context context,
            @Nullable final AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ScrubBar(
            @NonNull final Context context,
            @Nullable final AttributeSet attrs,
            final int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        mEnabledProgressColor = getColorAccent(context);

        final float density = context.getResources().getDisplayMetrics().density;
        mTrackWidth = TRACK_WIDTH_DP * density;
        mTrackWidthHalf = mTrackWidth / 2;
        mSmallThumbRadius = SMALL_THUMB_RADIUS_DP * density;
        mLargeThumbRadius = LARGE_THUMB_RADIUS_DP * density;
        mScrubUnitLength = (int) (SCRUB_UNIT_LENGTH_DP * density + 0.5f);
        mPaint = new Paint();
        mPaint.setStrokeWidth(mTrackWidth);
        mPaint.setAntiAlias(true);
        mProgressColor = isEnabled() ? mEnabledProgressColor : mDisabledProgressColor;
        mTrackColor = isEnabled() ? mEnabledTrackColor : mDisabledTrackColor;
        mSectionColor = isEnabled() ? mEnabledSectionColor : mDisabledSectionColor;
    }

    private static int getColorAccent(@NonNull final Context context) {
        final int colorAttr = getColorAccentId(context);
        final TypedValue outValue = new TypedValue();
        context.getTheme().resolveAttribute(colorAttr, outValue, true);
        return outValue.data;
    }

    private static int getColorAccentId(@NonNull final Context context) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            return android.R.attr.colorAccent;
        }
        return context.getResources().getIdentifier("colorAccent", "attr", context.getPackageName());
    }

    public void setProgressColor(@ColorInt final int color) {
        mEnabledProgressColor = color;
        mProgressColor = isEnabled() ? mEnabledProgressColor : mDisabledProgressColor;
        invalidate();
    }

    public void setTrackColor(@ColorInt final int color) {
        mEnabledTrackColor = color;
        mTrackColor = isEnabled() ? mEnabledTrackColor : mDisabledTrackColor;
        invalidate();
    }

    public void setDisabledProgressColor(@ColorInt final int color) {
        mDisabledProgressColor = color;
        mProgressColor = isEnabled() ? mEnabledProgressColor : mDisabledProgressColor;
        invalidate();
    }

    public void setDisabledTrackColor(@ColorInt final int color) {
        mDisabledTrackColor = color;
        mTrackColor = isEnabled() ? mEnabledTrackColor : mDisabledTrackColor;
        invalidate();
    }

    public void setSectionColor(@ColorInt final int color) {
        mEnabledSectionColor = color;
        mSectionColor = isEnabled() ? mEnabledSectionColor : mDisabledSectionColor;
        invalidate();
    }

    public void setDisabledSectionColor(@ColorInt final int color) {
        mDisabledSectionColor = color;
        mSectionColor = isEnabled() ? mEnabledSectionColor : mDisabledSectionColor;
        invalidate();
    }

    public void setScrubUnitLength(final int length) {
        mScrubUnitLength = length;
    }

    public void setTopBackgroundColor(@ColorInt final int color) {
        mTopBackgroundColor = color;
        invalidate();
    }

    public void setBottomBackgroundColor(@ColorInt final int color) {
        mBottomBackgroundColor = color;
        invalidate();
    }

    private static class SavedState extends BaseSavedState {
        private int progress;
        private int max;
        private List<Integer> chapterList;

        SavedState(@NonNull final Parcelable superState) {
            super(superState);
        }

        private SavedState(@NonNull final Parcel in) {
            super(in);
            progress = in.readInt();
            max = in.readInt();
            final int n = in.readInt();
            if (n == 0) {
                chapterList = Collections.emptyList();
                return;
            }
            chapterList = new ArrayList<>(n);
            for (int i = 0; i < n; i++) {
                chapterList.add(in.readInt());
            }
        }

        @Override
        public void writeToParcel(
                @NonNull final Parcel out,
                final int flags) {
            super.writeToParcel(out, flags);
            out.writeInt(progress);
            out.writeInt(max);
            out.writeInt(chapterList.size());
            for (final int chapter : chapterList) {
                out.writeInt(chapter);
            }
        }

        public static final Parcelable.Creator<SavedState> CREATOR
                = new Parcelable.Creator<SavedState>() {
            public SavedState createFromParcel(@NonNull final Parcel in) {
                return new SavedState(in);
            }

            public SavedState[] newArray(int size) {
                return new SavedState[size];
            }
        };
    }

    @Override
    protected Parcelable onSaveInstanceState() {
        final Parcelable superState = super.onSaveInstanceState();
        final SavedState ss = new SavedState(superState);
        ss.progress = mProgress;
        ss.max = mMax;
        ss.chapterList = mChapterList;
        return ss;
    }

    @Override
    protected void onRestoreInstanceState(@NonNull final Parcelable state) {
        final SavedState ss = (SavedState) state;
        super.onRestoreInstanceState(ss.getSuperState());
        setMax(ss.max);
        setProgress(ss.progress);
        setChapterList(ss.chapterList);
    }

    @Override
    public void setEnabled(final boolean enabled) {
        super.setEnabled(enabled);
        mProgressColor = enabled ? mEnabledProgressColor : mDisabledProgressColor;
        mTrackColor = enabled ? mEnabledTrackColor : mDisabledTrackColor;
        mSectionColor = enabled ? mEnabledSectionColor : mDisabledSectionColor;
    }

    @Override
    protected void onMeasure(
            final int widthMeasureSpec,
            final int heightMeasureSpec) {
        final int w = (int) (mTrackWidth + getPaddingLeft() + getPaddingRight());
        final int h = (int) (mTrackWidth + getPaddingTop() + getPaddingBottom());
        setMeasuredDimension(
                resolveSizeAndState(w, widthMeasureSpec, 0),
                resolveSizeAndState(h, heightMeasureSpec, 0));
    }

    @Override
    protected void onDraw(final Canvas canvas) {
        final int left = getPaddingLeft();
        final int right = getPaddingRight();
        final int top = getPaddingTop();
        final int bottom = getPaddingBottom();
        final int width = getWidth();
        final int height = getHeight();
        final float areaWidth = width - left - right;
        final float areaHeight = height - top - bottom;
        final float cy = areaHeight / 2 + top;
        final float l = mTrackWidthHalf;

        canvas.save();
        if (mTopBackgroundColor != 0) {
            canvas.clipRect(0, 0, width, cy + l);
            canvas.drawColor(mTopBackgroundColor);
        }
        canvas.restore();
        canvas.save();
        if (mBottomBackgroundColor != 0) {
            canvas.clipRect(0, cy - l, width, height);
            canvas.drawColor(mBottomBackgroundColor);
        }
        canvas.restore();

        mPaint.setColor(mTrackColor);
        canvas.drawLine(left, cy, areaWidth + left, cy, mPaint);
        if (mMax == 0) {
            return;
        }

        mPaint.setColor(mProgressColor);
        final float cx = mProgress * areaWidth / mMax + left;
        canvas.drawLine(left, cy, cx, cy, mPaint);
        final float radius = mDragging ? mLargeThumbRadius : mSmallThumbRadius;
        canvas.drawCircle(cx, cy, radius, mPaint);

        mPaint.setColor(mSectionColor);
        for (final int chapter : mChapterList) {
            final float sx = chapter * areaWidth / mMax + left;
            canvas.drawLine(sx - l, cy, sx + l, cy, mPaint);
        }
    }

    @Override
    public boolean onTouchEvent(@NonNull final MotionEvent event) {
        if (!isEnabled()) {
            return false;
        }
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                mDragging = true;
                onTouchStart(event);
                break;
            case MotionEvent.ACTION_MOVE:
                if (!mDragging) {
                    break;
                }
                onTouchMove(event);
                break;
            case MotionEvent.ACTION_CANCEL:
            case MotionEvent.ACTION_UP:
                if (!mDragging) {
                    break;
                }
                mDragging = false;
                onTouchEnd(event);
                break;
            default:
                break;
        }
        return true;
    }

    private void onTouchStart(@NonNull final MotionEvent event) {
        mStartX = event.getX();
        mStartY = event.getY();
        final int progress = getProgressByPosition(event.getX());
        mBaseProgress = progress;
        mAccuracyRank = 0;
        setProgressInternal(progress, true);
        mScrubBarListener.onStartTrackingTouch(this);
        mScrubBarListener.onAccuracyChanged(this, ACCURACY_RANKS[mAccuracyRank]);
    }

    private void onTouchMove(@NonNull final MotionEvent event) {
        final float dx = (event.getX() - mStartX) * ACCURACY[mAccuracyRank];
        final int progressDiff = getProgressByDistance(dx);
        setProgressInternal(progressDiff + mBaseProgress, true);

        final int distance = Math.abs((int) (event.getY() - mStartY));
        final int rank = clamp(distance / mScrubUnitLength, 0, RANK_MAX);
        if (mAccuracyRank != rank) {
            mAccuracyRank = rank;
            mBaseProgress = mProgress;
            mStartX = event.getX();
            mScrubBarListener.onAccuracyChanged(this, ACCURACY_RANKS[rank]);
        }
    }

    private void onTouchEnd(@NonNull final MotionEvent event) {
        onTouchMove(event);
        mScrubBarListener.onStopTrackingTouch(this);
        invalidate();
    }

    private int getProgressByPosition(final float x) {
        final int left = getPaddingLeft();
        final int right = getPaddingRight();
        final int width = getWidth() - left - right;
        if (width == 0) {
            return 0;
        }
        return (int) ((x - left) * mMax / width);
    }

    private int getProgressByDistance(final float dx) {
        final int left = getPaddingLeft();
        final int right = getPaddingRight();
        final int width = getWidth() - left - right;
        if (width == 0) {
            return 0;
        }
        return (int) (dx * mMax / width);
    }

    public void setChapterList(@Nullable final List<Integer> chapterList) {
        mChapterList = chapterList != null ? chapterList : Collections.emptyList();
        invalidate();
    }

    @NonNull
    public List<Integer> getChapterList() {
        return mChapterList;
    }

    public void setMax(int max) {
        max = Math.max(0, max);
        if (max == mMax) {
            return;
        }
        mMax = max;
        mProgress = Math.min(mProgress, mMax);
        invalidate();
    }

    public int getMax() {
        return mMax;
    }

    public void setProgress(final int progress) {
        setProgressInternal(progress, false);
    }

    public int getProgress() {
        return mProgress;
    }

    private void setProgressInternal(
            int progress,
            boolean fromUser) {
        progress = clamp(progress, 0, mMax);
        if (progress == mProgress) {
            return;
        }
        mProgress = progress;
        mScrubBarListener.onProgressChanged(this, progress, fromUser);
        invalidate();
    }

    public void setScrubBarListener(@Nullable final ScrubBarListener listener) {
        mScrubBarListener = listener != null ? listener : LISTENER;
    }

    private static int clamp(
            int value,
            int min,
            int max) {
        return Math.min(Math.max(value, min), max);
    }
}

File: mobile/src/main/java/net/mm2d/dmsexplorer/view/view/ScrubBar.kt
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.dmsexplorer.view.view

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator
import android.util.AttributeSet
import android.util.TypedValue
import android.view.MotionEvent
import android.view.View
import androidx.annotation.ColorInt
import androidx.annotation.Dimension
import androidx.annotation.IntDef
import kotlin.math.abs

/**
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 */
class ScrubBar @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0,
) : View(context, attrs, defStyleAttr) {
    private val paint: Paint
    private var scrubBarListener = LISTENER

    @Dimension(unit = Dimension.PX)
    private val trackWidth: Float

    @Dimension(unit = Dimension.PX)
    private val trackWidthHalf: Float

    @Dimension(unit = Dimension.PX)
    private val smallThumbRadius: Float

    @Dimension(unit = Dimension.PX)
    private val largeThumbRadius: Float

    @Dimension(unit = Dimension.PX)
    private var scrubUnitLength: Int = 0

    @ColorInt
    private var enabledProgressColor: Int = 0

    @ColorInt
    private var enabledTrackColor = DEFAULT_ENABLED_TRACK_COLOR

    @ColorInt
    private var enabledSectionColor = DEFAULT_ENABLED_SECTION_COLOR

    @ColorInt
    private var disabledProgressColor = DEFAULT_DISABLED_PROGRESS_COLOR

    @ColorInt
    private var disabledTrackColor = DEFAULT_DISABLED_TRACK_COLOR

    @ColorInt
    private var disabledSectionColor = DEFAULT_DISABLED_SECTION_COLOR

    @ColorInt
    private var progressColor: Int = 0

    @ColorInt
    private var trackColor: Int = 0

    @ColorInt
    private var sectionColor: Int = 0

    @ColorInt
    private var topBackgroundColor: Int = 0

    @ColorInt
    private var bottomBackgroundColor: Int = 0

    private var accuracyRank: Int = 0
    private var dragging: Boolean = false
    private var startX: Float = 0.toFloat()
    private var startY: Float = 0.toFloat()
    private var baseProgress: Int = 0

    private var _progress: Int = 0
    var progress: Int
        get() = _progress
        set(progress) = setProgressInternal(progress, false)

    var max: Int = 0
        set(value) {
            val max = value.coerceAtLeast(0)
            if (max == this.max) {
                return
            }
            field = max
            _progress = _progress.coerceAtMost(max)
            invalidate()
        }

    private var _chapterList: List<Int> = emptyList()
    var chapterList: List<Int>?
        get() = _chapterList
        set(chapterList) {
            _chapterList = chapterList ?: mutableListOf()
            invalidate()
        }

    interface ScrubBarListener {
        fun onProgressChanged(seekBar: ScrubBar, progress: Int, fromUser: Boolean)
        fun onStartTrackingTouch(seekBar: ScrubBar)
        fun onStopTrackingTouch(seekBar: ScrubBar)
        fun onAccuracyChanged(seekBar: ScrubBar, @Accuracy accuracy: Int)
    }

    init {
        enabledProgressColor = getColorAccent(context)
        val density = context.resources.displayMetrics.density
        trackWidth = TRACK_WIDTH_DP * density
        trackWidthHalf = trackWidth / 2
        smallThumbRadius = SMALL_THUMB_RADIUS_DP * density
        largeThumbRadius = LARGE_THUMB_RADIUS_DP * density
        scrubUnitLength = (SCRUB_UNIT_LENGTH_DP * density + 0.5f).toInt()
        paint = Paint()
        paint.strokeWidth = trackWidth
        paint.isAntiAlias = true
        progressColor = if (isEnabled) enabledProgressColor else disabledProgressColor
        trackColor = if (isEnabled) enabledTrackColor else disabledTrackColor
        sectionColor = if (isEnabled) enabledSectionColor else disabledSectionColor
    }

    fun setProgressColor(@ColorInt color: Int) {
        enabledProgressColor = color
        progressColor = if (isEnabled) enabledProgressColor else disabledProgressColor
        invalidate()
    }

    fun setTrackColor(@ColorInt color: Int) {
        enabledTrackColor = color
        trackColor = if (isEnabled) enabledTrackColor else disabledTrackColor
        invalidate()
    }

    fun setDisabledProgressColor(@ColorInt color: Int) {
        disabledProgressColor = color
        progressColor = if (isEnabled) enabledProgressColor else disabledProgressColor
        invalidate()
    }

    fun setDisabledTrackColor(@ColorInt color: Int) {
        disabledTrackColor = color
        trackColor = if (isEnabled) enabledTrackColor else disabledTrackColor
        invalidate()
    }

    fun setSectionColor(@ColorInt color: Int) {
        enabledSectionColor = color
        sectionColor = if (isEnabled) enabledSectionColor else disabledSectionColor
        invalidate()
    }

    fun setDisabledSectionColor(@ColorInt color: Int) {
        disabledSectionColor = color
        sectionColor = if (isEnabled) enabledSectionColor else disabledSectionColor
        invalidate()
    }

    fun setScrubUnitLength(length: Int) {
        scrubUnitLength = length
    }

    fun setTopBackgroundColor(@ColorInt color: Int) {
        topBackgroundColor = color
        invalidate()
    }

    fun setBottomBackgroundColor(@ColorInt color: Int) {
        bottomBackgroundColor = color
        invalidate()
    }

    private class SavedState : BaseSavedState {
        var progress: Int = 0
        var max: Int = 0
        var chapterList: List<Int> = emptyList()

        constructor(superState: Parcelable) : super(superState)

        private constructor(parcel: Parcel) : super(parcel) {
            progress = parcel.readInt()
            max = parcel.readInt()
            val n = parcel.readInt()
            if (n == 0) {
                chapterList = emptyList()
                return
            }
            val list = ArrayList<Int>(n)
            for (i in 0 until n) {
                list.add(parcel.readInt())
            }
            chapterList = list
        }

        override fun writeToParcel(
            out: Parcel,
            flags: Int,
        ) {
            super.writeToParcel(out, flags)
            out.writeInt(progress)
            out.writeInt(max)
            out.writeInt(chapterList.size)
            chapterList.forEach {
                out.writeInt(it)
            }
        }

        companion object CREATOR : Creator<SavedState> {
            override fun createFromParcel(`in`: Parcel): SavedState {
                return SavedState(`in`)
            }

            override fun newArray(size: Int): Array<SavedState?> {
                return arrayOfNulls(size)
            }
        }
    }

    override fun onSaveInstanceState(): Parcelable? {
        val superState = super.onSaveInstanceState() ?: return null
        val ss = SavedState(superState)
        ss.progress = _progress
        ss.max = max
        ss.chapterList = _chapterList
        return ss
    }

    override fun onRestoreInstanceState(state: Parcelable) {
        val ss = state as SavedState
        super.onRestoreInstanceState(ss.superState)
        max = ss.max
        progress = ss.progress
        chapterList = ss.chapterList
    }

    override fun setEnabled(enabled: Boolean) {
        super.setEnabled(enabled)
        progressColor = if (enabled) enabledProgressColor else disabledProgressColor
        trackColor = if (enabled) enabledTrackColor else disabledTrackColor
        sectionColor = if (enabled) enabledSectionColor else disabledSectionColor
    }

    override fun onMeasure(
        widthMeasureSpec: Int,
        heightMeasureSpec: Int,
    ) {
        val w = (trackWidth + paddingLeft.toFloat() + paddingRight.toFloat()).toInt()
        val h = (trackWidth + paddingTop.toFloat() + paddingBottom.toFloat()).toInt()
        setMeasuredDimension(
            resolveSizeAndState(w, widthMeasureSpec, 0),
            resolveSizeAndState(h, heightMeasureSpec, 0),
        )
    }

    override fun onDraw(canvas: Canvas) {
        val paddingLeft = paddingLeft
        val areaWidth = (width - paddingLeft - paddingRight).toFloat()
        val areaHeight = (height - paddingTop - paddingBottom).toFloat()
        val cy = areaHeight / 2 + paddingTop
        val l = trackWidthHalf

        if (topBackgroundColor != 0) {
            canvas.save()
            canvas.clipRect(0f, 0f, width.toFloat(), cy + l)
            canvas.drawColor(topBackgroundColor)
            canvas.restore()
        }
        if (bottomBackgroundColor != 0) {
            canvas.save()
            canvas.clipRect(0f, cy - l, width.toFloat(), height.toFloat())
            canvas.drawColor(bottomBackgroundColor)
            canvas.restore()
        }
        paint.color = trackColor
        canvas.drawLine(paddingLeft.toFloat(), cy, areaWidth + paddingLeft, cy, paint)
        if (max == 0) {
            return
        }

        paint.color = progressColor
        val cx = _progress * areaWidth / max + paddingLeft
        canvas.drawLine(paddingLeft.toFloat(), cy, cx, cy, paint)
        val radius = if (dragging) largeThumbRadius else smallThumbRadius
        canvas.drawCircle(cx, cy, radius, paint)

        paint.color = sectionColor
        _chapterList.forEach {
            val sx = it * areaWidth / max + paddingLeft
            canvas.drawLine(sx - l, cy, sx + l, cy, paint)
        }
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        if (!isEnabled) {
            return false
        }
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                dragging = true
                onTouchStart(event)
            }

            MotionEvent.ACTION_MOVE -> {
                if (dragging) {
                    onTouchMove(event)
                }
            }

            MotionEvent.ACTION_CANCEL, MotionEvent.ACTION_UP -> {
                if (dragging) {
                    dragging = false
                    onTouchEnd(event)
                }
            }

            else -> {
            }
        }
        return true
    }

    private fun onTouchStart(event: MotionEvent) {
        startX = event.x
        startY = event.y
        val progress = getProgressByPosition(event.x)
        baseProgress = progress
        accuracyRank = 0
        setProgressInternal(progress, true)
        scrubBarListener.onStartTrackingTouch(this)
        scrubBarListener.onAccuracyChanged(this, ACCURACY_RANKS[accuracyRank])
    }

    private fun onTouchMove(event: MotionEvent) {
        val dx = (event.x - startX) * ACCURACY[accuracyRank]
        val progressDiff = getProgressByDistance(dx)
        setProgressInternal(progressDiff + baseProgress, true)

        val distance = abs((event.y - startY).toInt())
        val rank = (distance / scrubUnitLength).coerceIn(0, RANK_MAX)
        if (accuracyRank != rank) {
            accuracyRank = rank
            baseProgress = _progress
            startX = event.x
            scrubBarListener.onAccuracyChanged(this, ACCURACY_RANKS[rank])
        }
    }

    private fun onTouchEnd(event: MotionEvent) {
        onTouchMove(event)
        scrubBarListener.onStopTrackingTouch(this)
        invalidate()
    }

    private fun getProgressByPosition(x: Float): Int {
        val left = paddingLeft
        val right = paddingRight
        val width = width - left - right
        return if (width == 0) 0 else ((x - left) * max / width).toInt()
    }

    private fun getProgressByDistance(dx: Float): Int {
        val left = paddingLeft
        val right = paddingRight
        val width = width - left - right
        return if (width == 0) 0 else (dx * max / width).toInt()
    }

    private fun setProgressInternal(
        progress: Int,
        fromUser: Boolean,
    ) {
        val newProgress = progress.coerceIn(0, max)
        if (newProgress == _progress) {
            return
        }
        _progress = newProgress
        scrubBarListener.onProgressChanged(this, newProgress, fromUser)
        invalidate()
    }

    fun setScrubBarListener(listener: ScrubBarListener?) {
        scrubBarListener = listener ?: LISTENER
    }

    @Retention(AnnotationRetention.SOURCE)
    @IntDef(ACCURACY_NORMAL, ACCURACY_HALF, ACCURACY_QUARTER)
    annotation class Accuracy

    companion object {
        @Dimension(unit = Dimension.DP)
        private val TRACK_WIDTH_DP = 3

        @Dimension(unit = Dimension.DP)
        private val SMALL_THUMB_RADIUS_DP = 4

        @Dimension(unit = Dimension.DP)
        private val LARGE_THUMB_RADIUS_DP = 8

        @Dimension(unit = Dimension.DP)
        private val SCRUB_UNIT_LENGTH_DP = 150

        @ColorInt
        private val DEFAULT_ENABLED_TRACK_COLOR = Color.argb(0x66, 0xff, 0xff, 0xff)

        @ColorInt
        private val DEFAULT_ENABLED_SECTION_COLOR = Color.argb(0xff, 0xff, 0xff, 0x0)

        @ColorInt
        private val DEFAULT_DISABLED_PROGRESS_COLOR = Color.argb(0xff, 0x80, 0x80, 0x80)

        @ColorInt
        private val DEFAULT_DISABLED_TRACK_COLOR = Color.argb(0x66, 0x60, 0x60, 0x60)

        @ColorInt
        private val DEFAULT_DISABLED_SECTION_COLOR = Color.argb(0x66, 0xff, 0xff, 0x0)

        const val ACCURACY_NORMAL = 0
        const val ACCURACY_HALF = 1
        const val ACCURACY_QUARTER = 2

        private val ACCURACY_RANKS = intArrayOf(
            ACCURACY_NORMAL,
            ACCURACY_HALF,
            ACCURACY_QUARTER,
        )
        private val ACCURACY = floatArrayOf(
            1.1f,
            0.5f,
            0.25f,
        )
        private val RANK_MAX = ACCURACY_RANKS.size - 1

        private val LISTENER: ScrubBarListener = object : ScrubBarListener {
            override fun onProgressChanged(
                seekBar: ScrubBar,
                progress: Int,
                fromUser: Boolean,
            ) = Unit

            override fun onStartTrackingTouch(seekBar: ScrubBar) = Unit
            override fun onStopTrackingTouch(seekBar: ScrubBar) = Unit
            override fun onAccuracyChanged(seekBar: ScrubBar, @Accuracy accuracy: Int) = Unit
        }

        private fun getColorAccent(context: Context): Int {
            val colorAttr = android.R.attr.colorAccent
            val outValue = TypedValue()
            context.theme.resolveAttribute(colorAttr, outValue, true)
            return outValue.data
        }
    }
}
